viewing_party/party.py

```diff
@@ -1,23 +1,142 @@
+        if movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+            return user_data
+    return user_data
+        
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    sum = 0.0
+    if not user_data["watched"]:
+        return sum
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+    average = sum / len(user_data["watched"])
+    return average
+
+def get_most_watched_genre(user_data):
+    genre_frequency_dict = {}
+    if not user_data["watched"]:
+        return None
+    for movie in user_data["watched"]:
+        if not movie["genre"] in genre_frequency_dict.keys():
+            genre_frequency_dict[movie["genre"]]= 1
+        else:
+            genre_frequency_dict[movie["genre"]] += 1
+    
+    most_watched_genre = max(genre_frequency_dict, key = genre_frequency_dict.get)
+    return most_watched_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            friend_watched_list.append(friend_movie["title"])
+
+    user_set = set(user_watched_list)
+    friend_set = set(friend_watched_list)
+    unique_set = user_set - friend_set
+
+    unique_list = []
+    for title in unique_set:
+        for movie in user_data["watched"]:
+            if title == movie["title"]:
+                unique_list.append(movie)
+    return unique_list
+
+def get_friends_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            friend_watched_list.append(friend_movie["title"])
+
+    user_set = set(user_watched_list)
+    friend_set = set(friend_watched_list)
+    unique_set = friend_set - user_set
+
+    unique_list = []
+    for title in unique_set:
+        for friend in user_data["friends"]:
+            for friend_movie in friend["watched"]:
+                if title == friend_movie["title"] and not (friend_movie in unique_list):
+                    unique_list.append(friend_movie)
+    return unique_list  
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    friend_unique = get_friends_unique_watched(user_data)
+    user_subscription_list = user_data["subscriptions"]

```

Comment:
This is a good idea to create a list that points to the user_data{"subscriptions"} list! However, later on, we aren't adding to or removing from the list in any way. Instead we are simply checking to see if a specific value appears in the list. Therefore, we don't necessarily need to make a new variable when we can reference the list specifically. On line 116, you use "user_subscription_list" but it could simply be replaced with user_data["subscriptions"] which avoids creating a new variable. If that is easier for you though, please ignore this advice!